content/en/blog/_posts/2021-12-15-StatefulSet-PVC-Auto-Deletion.md

```diff
@@ -0,0 +1,42 @@
+---
+layout: blog
+title: 'StatefulSet PVC Auto-Deletion'
+date: 2021-12-15
+slug: statefulset-pvc-auto-deletion
+---
+
+**Author:** Matthew Cary (Google)
+
+Kubernetes v1.23 introduced a new, alpha-level  policy for [StatefulSets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) that controls the lifetime of [PersistentVolumeClaims](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) (PVCs) generated from the StatefulSet spec template for cases when they should be deleted automatically when the StatefulSet is deleted or pods in the StatefulSet are scaled down.
+
+## What problem does this solve?
+A StatefulSet spec can include  Pod and PVC templates. When a replica is first created, the Kubernetes control plane creates a PVC for that replica if one does not already exist. The behavior before Kubernetes v1.23 was that the control plane never cleaned up the PVCs created for StatefulSets - this was left up to the cluster administrator, or to some add-on automation that you’d have to find, check suitability, and deploy. The common pattern for managing PVCs, either manually or through tools such as Helm, is that the PVCs are tracked by the tool that manages them, with explicit lifecycle. Workflows that use StatefulSets must determine on their own what PVCs are created by a StatefulSet and what their lifecycle should be.
+
+Before this new feature, when a StatefulSet-managed replica disappears, either because the StatefulSet is reducing its replica count, or because its StatefulSet is deleted, the PVC and its backing volume remains and must be manually deleted. While this behavior is appropriate when the data is critical, in many cases the persistent data in these PVCs is either temporary, or can be reconstructed from another source. In those cases, PVCs and their backing volumes remaining after their StatefulSet or replicas have been deleted are not necessary, incur cost, and require manual cleanup. 
+
+## The new StatefulSet PVC retention policy
+
+If you enable the alpha feature, a StatefulSet spec includes a PersistentVolumeClaim retention policy. This is used to control if and when PVCs created from a StatefulSet’s volumeClaimTemplate are deleted. This first iteration of the retention policy contains two situations where PVCs may be deleted. The first situation is when the StatefulSet resource is deleted (which implies that all replicas are also deleted). This is controlled by the **whenDeleted** policy. The second situation, controlled by **whenScaled** is when the StatefulSet is scaled down, which removes some but not all of the replicas in a StatefulSet. In both cases the policy can either be Retain, where the corresponding PVCs are not touched, or Delete, which means that PVCs are deleted. The deletion is done with a normal [object deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/), so that, for example, all retention policies for the underlying PV are respected.
```

Comment:
```suggestion
If you enable the alpha feature, a StatefulSet spec includes a PersistentVolumeClaim retention policy.
This is used to control if and when PVCs created from a StatefulSet’s `volumeClaimTemplate` are deleted.
This first iteration of the retention policy contains two situations where PVCs may be deleted.

The first situation is when the StatefulSet resource is deleted (which implies that all replicas
are also deleted). This is controlled by the `whenDeleted` policy. The second situation, controlled
by `whenScaled` is when the StatefulSet is scaled down, which removes some but not all of the
replicas in a StatefulSet. In both cases the policy can either be `Retain`, where the corresponding
PVCs are not touched, or `Delete`, which means that PVCs are deleted. The deletion is done with
a normal [object deletion](/docs/concepts/architecture/garbage-collection/), so that, for example,
all retention policies for the underlying PV are respected.
```